### PR TITLE
   Avoid print icons when running on Windows OS

### LINKED ISSUE #14219: Windows - ? instead of codestarts icons when generating project
Running https://quarkus.io/guides/getting-started#bootstrapping-the-project on Windows 10

There are ? instead of codestarts icons when generating project, tried with cmd and poweshell

Screenshot 2021-01-11 11 20 08

Screenshot 2021-01-11 11 21 35

Expected to see something like this:
Screenshot 2021-01-11 at 11 50 53

### PR DESCRIPTION
**Pull request by ejba**
Closes #14219

### CODE REVIEW

File: independent-projects/tools/codestarts/src/main/java/io/quarkus/devtools/codestarts/core/CodestartProjectGeneration.java [hunk 1]
```diff
@@ -17,6 +17,8 @@
 
 final class CodestartProjectGeneration {
 
+    private static final boolean IS_WINDOWS_OS = System.getProperty("os.name").toLowerCase().contains("win");
```
Comment (gastaldi) [inline independent-projects/tools/codestarts/src/main/java/io/quarkus/devtools/codestarts/core/CodestartProjectGeneration.java:20]: What if the OS is "Darwin"?

There is a `smallrye-common-os` library that implements that easily

File: independent-projects/tools/codestarts/src/main/java/io/quarkus/devtools/codestarts/core/CodestartProjectGeneration.java [hunk 2]
```diff
@@ -47,11 +48,19 @@ static void generateProject(final CodestartProjectDefinition projectDefinition,
...
-                .collect(Collectors.joining("\n")));
+        if (!OS.WINDOWS.isCurrent()) {
+            log.info(projectDefinition.getCodestarts().stream()
+                    .map(c -> c.getType().getIcon() + " "
+                            + c.getName())
+                    .collect(Collectors.joining("\n")));
+        } else {
+            log.info(projectDefinition.getCodestarts().stream()
+                    .map(c -> c.getType().getAlternativeIcon() + " " + c.getName())
+                    .collect(Collectors.joining("\n")));
```
Comment (gastaldi) [inline independent-projects/tools/codestarts/src/main/java/io/quarkus/devtools/codestarts/core/CodestartProjectGeneration.java:61]: Shouldn't this be `\r\n`? I wonder if we could use `System.lineSeparator()` instead in both statements

File: independent-projects/tools/codestarts/src/main/java/io/quarkus/devtools/codestarts/core/CodestartProjectGeneration.java [hunk 3]
```diff
@@ -47,11 +48,19 @@ static void generateProject(final CodestartProjectDefinition projectDefinition,
             processor.process(codestart);
         }
         processor.writeFiles();
+
         log.info("\napplying codestarts...");
-        log.info(projectDefinition.getCodestarts().stream()
-                .map(c -> c.getType().getIcon() + " "
-                        + c.getName())
-                .collect(Collectors.joining("\n")));
+        if (!OS.WINDOWS.isCurrent()) {
```
Comment (gastaldi) [inline independent-projects/tools/codestarts/src/main/java/io/quarkus/devtools/codestarts/core/CodestartProjectGeneration.java:53]: The code is easier to read if you invert this condition

Comment (ia3andy) [reply independent-projects/tools/codestarts/src/main/java/io/quarkus/devtools/codestarts/core/CodestartProjectGeneration.java]: Well it's maybe easier to ready if the condition is in the `getIcon()` no?

**Review by gastaldi — APPROVED** [review]
LGTM, but I'll let @ia3andy have the final word ;)

Good job!

File: independent-projects/tools/codestarts/src/main/java/io/quarkus/devtools/codestarts/core/CodestartProjectGeneration.java (hunk 2 above)
Comment (ejba) [reply independent-projects/tools/codestarts/src/main/java/io/quarkus/devtools/codestarts/core/CodestartProjectGeneration.java]: Done

File: independent-projects/tools/codestarts/src/main/java/io/quarkus/devtools/codestarts/core/CodestartProjectGeneration.java (hunk 3 above)
Comment (gastaldi) [reply independent-projects/tools/codestarts/src/main/java/io/quarkus/devtools/codestarts/core/CodestartProjectGeneration.java]: Sure, that would make the `getAlternativeIcon` no longer necessary

**Review by ia3andy — CHANGES_REQUESTED** [review]
Well, I think I would prefer having only `getIcon` (no `getAlternateIcon`), and have the switch inside the `getIcon` so that anyone calling it won't have any OS problem.. wdyt?

**Comment by ia3andy** [comment]
We shouldn't forget that we also have wrong display for those: https://github.com/quarkusio/quarkus/blob/master/independent-projects/tools/message-writer/src/main/java/io/quarkus/devtools/messagewriter/MessageIcons.java

File: independent-projects/tools/codestarts/src/main/java/io/quarkus/devtools/codestarts/core/CodestartProjectGeneration.java (hunk 3 above)
Comment (ejba) [reply independent-projects/tools/codestarts/src/main/java/io/quarkus/devtools/codestarts/core/CodestartProjectGeneration.java]: done

**Review by ia3andy — APPROVED** [review]
Nice :)

**Comment by gastaldi** [comment]
> We shouldn't forget that we also have wrong display for those: [https://github.com/quarkusio/quarkus/blob/master/independent-projects/tools/message-writer/src/main/java/io/quarkus/devtools/messagewriter/MessageIcons.java](https://github.com/quarkusio/quarkus/blob/master/independent-projects/tools/message-writer/src/main/java/io/quarkus/devtools/messagewriter/MessageIcons.java?rgh-link-date=2021-02-02T09%3A51%3A48Z)

Ah nice catch. Perhaps those need to be refactored as methods? It would be nice if those constants were not public

File: independent-projects/tools/message-writer/src/main/java/io/quarkus/devtools/messagewriter/MessageIcons.java [hunk 4]
```diff
@@ -1,14 +1,23 @@
...
 
-public final class MessageIcons {
+import io.smallrye.common.os.OS;
 
-    public static final String OK_ICON = "\u2705";
-    public static final String NOK_ICON = "\u274c";
-    public static final String NOOP_ICON = "\uD83D\uDC4D";
-    public static final String WARN_ICON = "\uD83D\uDD25";
-    public static final String ERROR_ICON = "\u2757";
+public enum MessageIcons {
```
Comment (gastaldi) [inline independent-projects/tools/message-writer/src/main/java/io/quarkus/devtools/messagewriter/MessageIcons.java:5]: Nice!

**Comment by ia3andy** [comment]
@ejba could you give a screenshot of the output when creating a project on Windows?

File: independent-projects/tools/message-writer/src/main/java/io/quarkus/devtools/messagewriter/MessageIcons.java [hunk 5]
```diff
@@ -1,14 +1,23 @@
...
+    WARN_ICON("\uD83D\uDD25"),
+    ERROR_ICON("\u2757");
+
+    String icon;
+
+    MessageIcons(String icon) {
+        this.icon = icon;
+    }
+
+    public String getIcon() {
```
Comment (gastaldi) [inline independent-projects/tools/message-writer/src/main/java/io/quarkus/devtools/messagewriter/MessageIcons.java:19]: Also make sure to override `toString()`, for cases where the constants here are concatenated into Strings

File: independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/commands/handlers/RemoveExtensionsCommandHandler.java [hunk 6]
```diff
@@ -44,7 +44,7 @@ public QuarkusCommandOutcome execute(QuarkusCommandInvocation invocation) throws
                 final UninstallResult result = extensionManager.uninstall(keys);
                 result.getUninstalled()
                         .forEach(a -> invocation.log()
-                                .info(MessageIcons.OK_ICON + " Extension " + a.getGroupId() + ":" + a.getArtifactId()
+                                .info(MessageIcons.OK_ICON.getIcon() + " Extension " + a.getGroupId() + ":" + a.getArtifactId()
```
Comment (gastaldi) [inline independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/commands/handlers/RemoveExtensionsCommandHandler.java:47]: Overriding toString() in the enum would avoid this change

File: independent-projects/tools/message-writer/src/main/java/io/quarkus/devtools/messagewriter/MessageIcons.java (hunk 5 above)
Comment (ejba) [reply independent-projects/tools/message-writer/src/main/java/io/quarkus/devtools/messagewriter/MessageIcons.java:19]: Done

File: independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/commands/handlers/RemoveExtensionsCommandHandler.java (hunk 6 above)
Comment (ejba) [reply independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/commands/handlers/RemoveExtensionsCommandHandler.java]: Done

**Comment by ejba** [comment]
@ia3andy I have to launch a VM for that, it would take time.

**Comment by ia3andy** [comment]
> @ia3andy I have to launch a VM for that, it would take time.

No worries, we may have them in the CI ;-)

**Comment by ejba** [comment]
It didn't work out. Should be use an asterisk instead?

**Comment by gastaldi** [comment]
> It didn't work out. Should be use an asterisk instead?

Sure, that makes sense

**Comment by ejba** [comment]
How do we set up the encoding? Shouldn't it be specified on DefaultMessageWriter?

**Comment by ia3andy** [comment]
For `MessageIcons`, I think we could have text replacement for Windows `OK_ICON = "[OK]"`...

Also, I think changing `MessageIcons` class will break the API and can't be backported (cc @aloubyansky) but I am not sure?

For the codestart type icons, you may use `-` or `>>` or whatever similar..

**Comment by aloubyansky** [comment]
Yes, it will break the backward compatibility. If it's still the desired change we can apply it at the time we extract the dev tools project into its own repo.

**Comment by ejba** [comment]
Can you trigger CI again? It's failing for other reasons not related with the PR.

**Comment by ia3andy** [comment]
> Can you trigger CI again? It's failing for other reasons not related with the PR.

done

**Comment by ia3andy** [comment]
cool thanks:

File: independent-projects/tools/message-writer/src/main/java/io/quarkus/devtools/messagewriter/MessageIcons.java [hunk 7]
```diff
@@ -1,14 +1,29 @@
...
-    public static final String NOK_ICON = "\u274c";
-    public static final String NOOP_ICON = "\uD83D\uDC4D";
-    public static final String WARN_ICON = "\uD83D\uDD25";
-    public static final String ERROR_ICON = "\u2757";
+public enum MessageIcons {
 
-    private MessageIcons() {
+    OK_ICON("\u2705", "[OK]"),
+    NOK_ICON("\u274c", "[NOK]"),
+    NOOP_ICON("\uD83D\uDC4D", "[NOOP]"),
```
Comment (ia3andy) [inline independent-projects/tools/message-writer/src/main/java/io/quarkus/devtools/messagewriter/MessageIcons.java:9]:
```suggestion
    NOOP_ICON("\uD83D\uDC4D", ":)"),
```

Comment (ia3andy) [inline independent-projects/tools/message-writer/src/main/java/io/quarkus/devtools/messagewriter/MessageIcons.java:9]: or a bit less funny
```suggestion
    NOOP_ICON("\uD83D\uDC4D", ""),
```

Comment (gastaldi) [reply independent-projects/tools/message-writer/src/main/java/io/quarkus/devtools/messagewriter/MessageIcons.java:9]: If using smiles, use `:(` for NOK and `;)` for OK 😉

Comment (ejba) [reply independent-projects/tools/message-writer/src/main/java/io/quarkus/devtools/messagewriter/MessageIcons.java:9]: Please, provide a list with the values you want for each case.

Comment (gsmet) [reply independent-projects/tools/message-writer/src/main/java/io/quarkus/devtools/messagewriter/MessageIcons.java:9]: Seriously, I don't think we should use smiley for that. It's really unclear.

Comment (gsmet) [reply independent-projects/tools/message-writer/src/main/java/io/quarkus/devtools/messagewriter/MessageIcons.java:9]: And also it's harder to parse if you want to look for errors. Let it be `[SUCCESS]` `[FAILURE]` as in Maven. That would be a lot clearer and easier to manipulate if needed.

Comment (ejba) [reply independent-projects/tools/message-writer/src/main/java/io/quarkus/devtools/messagewriter/MessageIcons.java]: Let me see if I got this straight. @gsmet So you prefer sending `[SUCCESS]` `[FAILURE]` `[WARNING]` `[ERROR]` instead of all these icons, right?

File: independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/commands/handlers/CreateProjectCommandHandler.java [hunk 8]
```diff
@@ -108,7 +107,7 @@ public QuarkusCommandOutcome execute(QuarkusCommandInvocation invocation) throws
             final CodestartProjectDefinition projectDefinition = catalog.createProject(input);
             projectDefinition.generate(invocation.getQuarkusProject().getProjectDirPath());
             invocation.log()
-                    .info("\n-----------\n" + MessageIcons.NOOP_ICON + " "
+                    .info("\n-----------\n" + " "
```
Comment (ia3andy) [inline independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/commands/handlers/CreateProjectCommandHandler.java:110]: Why did you remove this icon?, This was badly named but looked great:
image

**Comment by ia3andy** [comment]
@gsmet @gastaldi @ejba did we decide to also change the icons on Unix systems? I really liked the way it looked before with thumbs up and all.

**Comment by ejba** [comment]
@ia3andy We don't have a final decision I think. I did a second commit to be easy to revert the changes. :)

File: independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/commands/handlers/CreateProjectCommandHandler.java (hunk 8 above)
Comment (ia3andy) [reply independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/commands/handlers/CreateProjectCommandHandler.java:110]: We should add a new icon named `SUCCESS_ICON` with the thumbs up and [SUCCESS] on Windows to use here and everywhere we used the NOOP to mean SUCCESS.

If there are some place `[NOOP]` is used a it should  (I am not sure there is), we should also choose a new unix icon for it.

**Comment by gsmet** [comment]
@ia3andy esthetics is one thing (and it's very personal, I personally don't like emojis, too aggressive for the eye) but using emojis for status report is IMHO counter productive as harder to parse and making things different for Unix and Windows is also a problem as you can't have a tool properly parsing things for both without having to deal with both cases (and you have to think about it).

I'm -1 for using emojis on status reports.

**Comment by ejba** [comment]
@gsmet What should I do with the enum `CodeStartType` then?

**Comment by ia3andy** [comment]
@ejba wait a bit before we settle on this ;-)

**Comment by ia3andy** [comment]
I think this should be decided in on the quarkus-dev mailing list. 

IMO the goal of those icons was purely esthetic and not meant to be parsed in the first place.., in the end maybe we should have both.

**Comment by ejba** [comment]
Having both ways raises not warn I believe, and both of you get what you want.

**Comment by ia3andy** [comment]
So, if @gsmet is ok, we could:
- For MessageIcons, use text AND icon when that compatible and only text when it's not
- For codestarts types icons, keep it as already done in the PR, they are really not meant to be parsed

Then I think we should rename `MessageIcons` to `MessageStatus`, and add a new `SUCCESS` status to replace wrongly used `NOOP`.

**Comment by gsmet** [comment]
Good for me!

**Comment by geoand** [comment]
What's the deal with this?

**Comment by ia3andy** [comment]
It's waiting for @ejba update.

**Comment by ejba** [comment]
@geoand done

**Comment by geoand** [comment]
@ia3andy mind doing a final check please and then merging if all is still OK?

Thanks

File: independent-projects/tools/message-writer/src/main/java/io/quarkus/devtools/messagewriter/MessageIcons.java [hunk 9]
```diff
@@ -1,14 +1,25 @@
...
-    public static final String NOK_ICON = "\u274c";
-    public static final String NOOP_ICON = "\uD83D\uDC4D";
-    public static final String WARN_ICON = "\uD83D\uDD25";
-    public static final String ERROR_ICON = "\u2757";
+public enum MessageIcons {
 
-    private MessageIcons() {
+    OK_ICON("\u2705", "[SUCCESS]"),
+    NOK_ICON("\u274c", "[FAILURE]"),
+    NOOP_ICON("\uD83D\uDC4D", "[FAILURE]"),
```
Comment (ia3andy) [inline independent-projects/tools/message-writer/src/main/java/io/quarkus/devtools/messagewriter/MessageIcons.java:9]:
```suggestion
    NOOP_ICON("\uD83D\uDC4D", ""),
```

Comment (ia3andy) [reply independent-projects/tools/message-writer/src/main/java/io/quarkus/devtools/messagewriter/MessageIcons.java:9]: NOOP means no operation but doesn't mean failure at all. I would put an empty string her.

**Review by ia3andy — CHANGES_REQUESTED** [review]

File: independent-projects/tools/message-writer/src/main/java/io/quarkus/devtools/messagewriter/MessageIcons.java (hunk 9 above)
Comment (ia3andy) [reply independent-projects/tools/message-writer/src/main/java/io/quarkus/devtools/messagewriter/MessageIcons.java:9]:
```
applying codestarts...
>> java
>> maven
>> quarkus
>> config-properties
>> dockerfiles
>> maven-wrapper
>> resteasy-example

-----------
[FAILURE] quarkus project has been successfully generated in:
--> target\extensions-test
```

:)

Comment (ia3andy) [reply independent-projects/tools/message-writer/src/main/java/io/quarkus/devtools/messagewriter/MessageIcons.java:9]: BTW I don't think for this message we should use a NOOP, it should be a OK.

**Review by ia3andy — APPROVED** [review]
Thanks a lot @ejba, it took a few rounds but most important is this is very nice now..

**Comment by ejba** [comment]
I did a rebase to trigger CI again. It's supposed to pass in all tests, right?

**Comment by geoand** [comment]
The failure has nothing to do with the PR, we can merge
